universal-application-tool-0.0.1/app/services/applicant/ApplicantServiceImpl.java

```diff
@@ -61,4 +99,59 @@ public ApplicantServiceImpl(
             },
             httpExecutionContext.current());
   }
+
+  /**
+   * Tries to perform an in-place update of {@link ApplicantData}. Returns error messages if it is
+   * impossible to perform the operation.
+   */
+  private ImmutableSet<String> stageUpdates(
+      Applicant applicant,
+      ProgramDefinition programDefinition,
+      long blockId,
+      ImmutableSet<Update> updates) {
+
+    Optional<BlockDefinition> blockDefinitionMaybe = programDefinition.getBlockDefinition(blockId);
+    if (blockDefinitionMaybe.isEmpty()) {
+      return ImmutableSet.of("Block not found.");
+    }
+    BlockDefinition blockDefinition = blockDefinitionMaybe.get();
+
+    ImmutableList<Path> updatePaths =
+        updates.stream().map(Update::path).collect(ImmutableList.toImmutableList());
+    if (!blockDefinition.hasPaths(updatePaths)) {
+      return ImmutableSet.of("Not all paths exist in this Block.");
+    }
+
+    try {
+      stageUpdates(applicant.getApplicantData(), blockDefinition, updates);
+    } catch (UnsupportedScalarTypeException e) {
+      return ImmutableSet.of("Unrecognized scalar type");
+    } catch (PathNotInBlockException e) {
+      return ImmutableSet.of("Update path does not exist.");
+    }
+
+    return ImmutableSet.of();
+  }
+
+  /** In-place update of applicant data. */
+  private void stageUpdates(
+      ApplicantData applicantData, BlockDefinition blockDefinition, ImmutableSet<Update> updates)
+      throws UnsupportedScalarTypeException, PathNotInBlockException {
+    for (Update update : updates) {
+      Optional<ScalarType> typeMaybe = blockDefinition.getScalarType(update.path());
+      if (typeMaybe.isEmpty()) {
+        throw new PathNotInBlockException(blockDefinition, update.path());
+      }
+      switch (typeMaybe.get()) {
+        case STRING:
+          applicantData.putString(update.path(), (String) update.value());
+          break;
+        case INT:
+          applicantData.putInteger(update.path(), (Integer) update.value());
```

Comment:
Issue (blocking): let's not cast if we can avoid it

Reply:
done